server/controllers/users/changePassword.js

```diff
@@ -1,3 +1,30 @@
-exports.changePassword = (req, res) => {
-  res.send('hello from changePassword file');
+const { hash, compare } = require('bcrypt');
+const { users: { changePassword, getUserById } } = require('../../database/queries');
+
+exports.changePassword = (req, res, next) => {
+  const { userId, oldPassword, newPassword } = req.body;
+  let userInfo;
+  getUserById(userId)
+    .then(({ rows }) => {
+      userInfo = rows;
+      return compare(oldPassword, (userInfo[0].password));
+    })
+    .then((checkOldPassword) => {
+      if (checkOldPassword) {
+        return compare(newPassword, (userInfo[0].password));
+      }
+      throw new Error('You entered a wrong password');
+    })
+    .then((checkNewPassword) => {
+      if (!checkNewPassword) {
+        hash(newPassword, 10);
+        return changePassword(userId, newPassword);
+      }
+      throw new Error('Your new password and old password are the same');
+    })
+    .then(() => res.send('Password is changed'))
+    .catch((err) => {
+      if (err.message) next({ code: 400, msg: err.message });
+      else next(err);
+    });
```

Comment:
it's not bad, but here is my suggestion on how to write it more concisely:
```
getUserById(userId).then(({ users }) => {
    if (users.length === 0) throw new Error(); // should return 404 here
    return Promise.all([
        compare(oldPassword, (users[0].password)),
        compare(newPassword, (users[0].password)),
    ])
}.then(([oldPasswordMatches, newPasswordMatches]) => {
    if (!oldPasswordMatches) throw new Error('You entered a wrong password')
    else if (newPasswordMatches) throw new Error('Your new password and old password are the same');
    
    hash(newPassword, 10);
    return changePassword(userId, newPassword);
}).then() // the rest is the same
```

Reply:
Thank you, I updated the code in another pull request #84

Reply:
@MaiUbeid may please review my last 2 commits on your branch
i did it with fatma
it doesn't make much sense that you did another pull request, would you mind why did you do that?

Reply:
ah, I made a mistake, you can put the `hash(newPassword, 10)` into the `Promise.all()`, then do `return changePassword()` in the first `then()`.
Just to be clear, this approach changes the execution flow. In your initial solution, the actions were executed sequentially, i.e. if the first check failed, the rest was not executed. Here, they are executed simultaneously, which improves performance for positive cases (no error), but makes it worse for negative cases (all of the actions in the Promise.all() will be executed always, even if one of them fails). So that is something to be considered, but in this particular case it's worth doing IMO.